lib/adf_diag.py

```diff
@@ -1017,66 +1028,135 @@ def derive_variables(self, vars_to_derive=None, ts_dir=None, overwrite=None):
-                        constit_files += elem
-                else:
-                    ermsg = "FSNT and FLNT were not both present; RESTOM cannot be calculated."
-                    ermsg += " Please remove RESTOM from diag_var_list or find the relevant CAM files."
-                    raise FileNotFoundError(ermsg)
-                # create new file name for RESTOM
-                derived_file = constit_files[0].replace("FLNT", "RESTOM")
+            print(f"\t - deriving time series for {var}")
+
+            #Check whether there are parts to derive from and if there is an associated equation
+            vres = res.get(var, {})
+            if "derivable_from" in vres:
+                constit_list = vres['derivable_from']
+            else:
+                print("WARNING: No constituents listed in defaults config file, moving on")
+                continue
+
+            constit_files = []
+            for constit in constit_list:
+                if glob.glob(os.path.join(ts_dir, f"*.{constit}.*.nc")):
+                    constit_files.append(glob.glob(os.path.join(ts_dir, f"*.{constit}.*"))[0])
+
+            #Check if all the constituent files were found
+            if len(constit_files) != len(constit_list):
+                ermsg = f"Not all constituent files present; {var} cannot be calculated."
+                ermsg += f" Please remove {var} from diag_var_list or find the relevant CAM files."
+                print(ermsg)
+            else:
+                #Open a new dataset with all the constituent files/variables
+                ds = xr.open_mfdataset(constit_files)
+
+                # create new file name for derived variable
+                derived_file = constit_files[0].replace(constit_list[0], var)
+                print(derived_file,"\n")
                 if Path(derived_file).is_file():
                     if overwrite:
                         Path(derived_file).unlink()
                     else:
                         print(
-                            f"[{__name__}] Warning: RESTOM file was found and overwrite is False. Will use existing file."
+                            f"[{__name__}] Warning: '{var}' file was found and overwrite is False. Will use existing file."
                         )
-                        continue
-                # append FSNT to the file containing FLNT
-                os.system(f"ncks -A -v FLNT {constit_files[0]} {constit_files[1]}")
-                # create new file with the difference of FLNT and FSNT
-                os.system(
-                    f"ncap2 -s 'RESTOM=(FSNT-FLNT)' {constit_files[1]} {derived_file}"
-                )
+                        return None
+
+                variables = {}
+                for i in constit_list:
+                    variables[i] = ds[constit_list[0]].dims
+
+                #Get coordinate values from the first constituent file
+                coords={'lat': ds[constit_list[0]].lat.values,
```

Comment:
Could this be changed to 
`coords = ds.coords`

That way we get the metadata in the coordinate variables, and we don't have to make assumptions about what coordinates are present. I'm thinking of a day when we can ingest data on unstructured grids.

Reply:
I tried this way, but when it gets to the aerosols (vertical levels) it seems to have a mismatch between coordinates and dimensions:

```ValueError: coordinate ilev has dimensions ('ilev',), but these are not a subset of the DataArray dimensions ('time', 'lev', 'lat', 'lon')```

---

```coords = ds.coords```

```
Coordinates:
  * lat      (lat) float64 -90.0 -89.06 -88.12 -87.17 ... 87.17 88.12 89.06 90.0
  * lon      (lon) float64 0.0 1.25 2.5 3.75 5.0 ... 355.0 356.3 357.5 358.8
  * lev      (lev) float64 0.008135 0.01834 0.03482 ... 983.2 991.2 997.5
  * ilev     (ilev) float64 0.004259 0.01201 0.02467 ... 987.4 995.1 1e+03
  * time     (time) object 1997-02-01 00:00:00 ... 2001-01-01 00:00:00
```

As per @nusbaume suggestion for setting the dims once:

```const_dims = ds_constits[constit_list[0]].dims```

```
('time', 'lev', 'lat', 'lon')
```

---

So I just grabbed the dimensions and coordinates from the first constituent dataArray

```
#Grab subset for first constituent for dimensions and coordinates
ds_const0 = ds[constit_list[0]]
```

```
coords = ds_const0.coords

KeysView(Coordinates:
  * lat      (lat) float64 -90.0 -89.06 -88.12 -87.17 ... 87.17 88.12 89.06 90.0
  * lon      (lon) float64 0.0 1.25 2.5 3.75 5.0 ... 355.0 356.3 357.5 358.8
  * lev      (lev) float64 0.008135 0.01834 0.03482 ... 983.2 991.2 997.5
  * time     (time) object 1997-02-01 00:00:00 ... 2001-01-01 00:00:00)
```

```
const_dims = ds_const0.dims
const_dims ('time', 'lev', 'lat', 'lon') 

```


It seemed to work, let me know if there is anything that stands out about going this way.